python/ml4ir/applications/ranking/model/losses/listwise_losses.py

```diff
@@ -32,7 +35,104 @@ def _loss_fn(y_true, y_pred):
             y_pred : [batch_size, num_classes]
             mask : [batch_size, num_classes]
             """
-            return cce(y_true, tf.math.multiply(y_pred, mask))
+
+            #Fixme
+            """
+            Queries with ties in the highest scores would have multiple one's in the 1-hot vector.
+            Queries with all zeros for y_true would have all ones as their 1-hot vector. 
+            """
+            if is_aux_loss:  # converting y-true to 1-hot for cce
+                y_true_1_hot = tf.equal(y_true, tf.expand_dims(tf.math.reduce_max(y_true, axis=1), axis=1))
+                y_true_1_hot = tf.cast(y_true_1_hot, dtype=tf.float32)
+                return cce(y_true_1_hot, tf.math.multiply(y_pred, mask))
+            else:
+                return cce(y_true, tf.math.multiply(y_pred, mask))
+
+        return _loss_fn
+
+    def get_final_activation_op(self, output_name):
+        """
+        Define a masked softmax activation function.
+        This is one of the simplest and most effective loss functions
+        for training ranking models with single click label.
+
+        Ref -> https://dl.acm.org/doi/10.1145/3341981.3344221
+
+        Parameters
+        ----------
+        output_name : str
+            Name of the output to apply softmax activation on
+
+        Returns
+        -------
+        function
+            Function to compute masked softmax
+
+        Notes
+        -----
+            Uses `mask` field to exclude padded records from contributing
+            to the softmax activation
+        """
+        softmax_op = layers.Softmax(axis=-1, name=output_name)
+
+        def masked_softmax(logits, mask):
+            """
+            NOTE:
+            Tried to manually compute softmax with tf operations,
+            but tf.keras.layers.Softmax() is more stable when working with
+            cross_entropy layers
+            """
+            logits = tf.where(
+                tf.equal(mask, tf.constant(1.0)), logits, tf.constant(tf.float32.min)
+            )
+
+            return softmax_op(logits)
+
+        return masked_softmax
+
+
+class BasicCrossEntropy(ListwiseLossBase):
+    def get_loss_fn(self, **kwargs):
+        """
+        Define a softmax cross entropy loss
+
+        Returns
+        -------
+        function
+            Function to compute softmax cross entropy loss
+
+        Notes
+        -----
+            Uses `mask` field to exclude padded records from contributing
+            to the loss
+        """
+        mask = kwargs.get("mask")
+        is_aux_loss = False
+        if kwargs.get("is_aux_loss"):
+            is_aux_loss = True
+        batch_size = kwargs.get("batch_size", 1)
+
+        def _loss_fn(y_true, y_pred):
+            """
+            Shapes
+            ------
+            y_true : [batch_size, num_classes]
```

Comment:
This loss expect y-true to be a prob dist. I explicitly apply a softmax for aux loss. I can enforce it for non aux loss as well

Reply:
You're converting it to a probability distribution, so you don't expect one at the input right?

Reply:
Both, y_pred and y_true should be prob dist. However, since the ranking pipeline already converts the predictions to a prob. dist before computing the loss, I'm explicitly just converting y_true to a prob dist.

Reply:
Yes, I understand. I wanted you to add that to the documentation :)